File: URI-1081.cpp
/*
* Problema: https://www.beecrowd.com.br/judge/pt/runs/code/28111113
* Presentation error (5%)
* 
* Autor: Alfredo
*/

#include <iostream>
#include<bits/stdc++.h>
#include <vector>
using namespace std;


int nivel = 0; //usado para fazer a identação

class Vertice{
    public:
    char cor; //b=branco / c=cinza / p=preto

    public:
    Vertice(){
        this->cor = 'b';
    }
};

class Grafo{
    public:
    int num_v; //numero de vertices
    int num_e; //numero de arestas
    bool **matAdj; //matriz adjacente
    vector<Vertice> listaVertices;

    public:
    Grafo (int v, int e){
        this->num_v=v;
        this->num_e=e;

        matAdj = new bool* [v]; // Create a dynamic array of pointers 
    
        for (int i=0; i<v; i++) // Create a row for every pointer 
        { 
            // Note : Rows may not be contiguous 
            matAdj[i] = new bool[v]; 
        
            // Inicializa todas as posições com falso;
            memset(matAdj[i], false, v*sizeof(bool)); 
        }
        alocaVertices();
    }
    void addAresta(int x, int y){ //adiciona nova aresta
        matAdj[x][y]=true;
    }
    void alocaVertices(){
        for(int i=0; i<num_v ; i++){
            Vertice vert;
            listaVertices.push_back(vert);
        }
    }
};

bool DFSVisit(Grafo *graf, int vert){ //retorna 1 caso exista ao menos 1 aresta neste vertice / retorna 1 caso não ache arestas nesse vertice
    bool ret=false;
    nivel++;
    graf->listaVertices[vert].cor='c'; //muda cor para cinza
    for(int i=0;i<graf->num_v;i++){ //procura qual vertice tem aresta com o vertice atual
        if(graf->matAdj[vert][i]==true){       //verifica se é aresta
            ret=true;
            for (int count=0; count<nivel; count++){ //identação
                cout << "  ";
            }
            cout << vert << "-" << i; 
            if(graf->listaVertices[i].cor=='b'){   //sempre q achar um vertice branco imprime o pathR e visita o proximo vertice
                cout << " pathR (G," << i << ")" << endl;
                DFSVisit(graf,i);
            }
            else{
                cout << endl; //somente termina a linha caso seja uma aresta de retorno
            }
        }
    }
    graf->listaVertices[vert].cor='p'; //pinta o vertice de preto ao terminar a visita
    nivel--;
    return ret;
}

int main() {

    int n; //numero de casos
	int v; //numero de vertices
	int e; //numero de arestas
    int x; //vertice de origem da aresta
    int y; //vertice de destino da aresta
    int r; // retorno da funçao dfsvisit (para tratar o caso similar ao do vertice 9 no caso 1 que imprime uma linha a mais por não ter arestas e ser visitado)

    cin >> n;

    for (int l=0; l<n;l++){   //roda para n casos
        cin >> v;
        cin >> e;
        Grafo g(v,e);
        for (int p=0; p<e ; p++){ //le as arestas e monta o grafo
            cin >> x; 
            cin >> y;
            g.addAresta(x,y);
        }
        cout << "Caso " << l+1 << ":" << endl;
        for (int i=0; i<v; i++){ //varre todos os vertices do grafo e se for branco faz a DFS (por linha)
            if(g.listaVertices[i].cor=='b'){
                r = DFSVisit(&g,i); //visita o vertice
                if(r){ //tratando o caso do vertice que não tem nenhuma aresta (para não imprimir uma linha a mais)
                    cout << endl; //pula uma linha quando muda de componente do grafo
                }
            }
            
        }
    }
    return 0;
}

File: URI-1081.rb
# 1081 - DFSr - Depth Hierarchy
# Challenge Link: https://www.beecrowd.com.br/judge/en/problems/view/1081
# Score: Wrong answer (5%)
# Alfredo Diani

class Vertice
    attr_accessor :cor

    def initialize
        @cor = 'b'
    end
end

class Grafo
    attr_accessor :num_v, :num_e, :matAdj, :listaVertices

    def initialize(v, e)
        @num_v = v
        @num_e = e
        @matAdj = Array.new(v) {Array.new(v, false)}
        @listaVertices = []
        aloca_vertices
    end

    def add_aresta(x, y)
        @matAdj[x][y] = true
    end

    def aloca_vertices
        @listaVertices = []
        (0...@num_v).each do |i|
            @listaVertices << Vertice.new
        end
    end
    
    def reset_vertices
        @listaVertices.each {|vert| vert.cor = 'b'}
    end
end

def DFS_visit(graf, vert, nivel, path_count, starting_vert)
    ret = false
    nivel += 1
    graf.listaVertices[vert].cor = 'c'
    (0...graf.num_v).each do |i|
        if graf.matAdj[vert][i]
            ret = true
            path_count += 1
            nivel.times {print "  "}
            if graf.listaVertices[i].cor != 'b'
                print "#{vert}-#{i}"
            else
                print "#{vert}-#{i} pathR(G,#{i})"
            end
            puts
            if graf.listaVertices[i].cor == 'b'
                DFS_visit(graf, i, nivel, path_count, starting_vert)
            end
        end
    end
    if path_count == graf.num_e
        graf.listaVertices[vert].cor = 'b' if vert != starting_vert
        path_count = 0
        return
    end
    ret
end

# Example usage
num_tests = gets.to_i
num_tests.times do |i|
    puts "Caso #{i+1}:"
    v, e = gets.split.map(&:to_i)
    graf = Grafo.new(v, e)
    e.times do
        x, y = gets.split.map(&:to_i)
        graf.add_aresta(x, y)
    end
    nivel = 0
    path_count = 0
    (0...graf.num_v).each do |j|
        if graf.listaVertices[j].cor == 'b'
            path_count = 0
            if DFS_visit(graf, j, nivel, path_count, j)
                puts
            end
        end
    end
    graf.reset_vertices
    puts if i != num_tests - 1
end

